File: csharp/sum-of-multiples/SumOfMultiples.cs
using System;
using System.Collections.Generic;
using System.Linq;

public static class SumOfMultiples
{
    public static int Sum(IEnumerable<int> multiples, int max)
    {
        return multiples
            .Where(n => n != 0)
            .SelectMany(n =>
                Enumerable
                .Range(0, max)
                .Where(x =>
                    x % n == 0
                )
            )
            .Distinct()
            .Sum();
    }
}

File: fsharp/sum-of-multiples/SumOfMultiples.fs
module SumOfMultiples

let sum (numbers: int list) (upperBound: int): int =
    let addItem upperBound itemBase =
        match itemBase with
        | 0 -> []
        | _ -> [itemBase..itemBase..upperBound]

    let addItemBakedUpperBound = addItem (upperBound - 1)

    numbers
    |> List.collect addItemBakedUpperBound
    |> List.distinct
    |> List.sum